### PR TITLE
[New] `DirectionProvider`: add `lang` prop

### PR DESCRIPTION
**Pull request by jpnelson**
This adds a `lang` prop to `DirectionProvider` and `WithDirectionProvider`

Generally when we want to switch direction, it's also the case that we want to switch languages. In addition, it'd be nice to be able to use the `:lang` pseudo class: https://developer.mozilla.org/en-US/docs/Web/CSS/:lang

### CODE REVIEW

**Comment by ljharb** [comment]
@jpnelson heads up that CI is broken on this package right now in node 4.

File: src/DirectionProvider.jsx [hunk 1]
```diff
@@ -45,10 +46,10 @@ export default class DirectionProvider extends React.Component {
   }
 
   render() {
-    const { children, direction, inline } = this.props;
+    const { children, direction, inline, lang } = this.props;
     const Tag = inline ? 'span' : 'div';
     return (
-      <Tag dir={direction}>
+      <Tag dir={direction} lang={lang}>
```
Comment (ljharb) [inline src/DirectionProvider.jsx]:
```suggestion
      <Tag dir={direction} ...(lang && { lang })>
```
that way there's no `lang` prop when `lang` isn't provided?

File: src/AutoDirectionProvider.jsx [hunk 2]
```diff
@@ -30,6 +32,7 @@ function AutoDirectionProvider({
     <DirectionProvider
       direction={dir}
       inline={inline}
+      lang={lang}
```
Comment (ljharb) [inline src/AutoDirectionProvider.jsx]:
```suggestion
      ...(lang && { lang })
```

**Review by ljharb — COMMENTED** [review]
Should there be a propType that restricts it to a valid language code?

File: README.md [hunk 3]
```diff
@@ -63,6 +63,22 @@ import DirectionProvider, { DIRECTIONS } from 'react-with-direction/dist/Directi
 </DirectionProvider>
 ```
 
+To set the `lang` attribute on the wrapping element, provide the `lang` prop to `DirectionProvider`.
+
+Usage example:
+
+```js
```
Comment (lencioni) [inline README.md]:
```suggestion
```jsx
```

File: src/DirectionProvider.jsx (hunk 1 above)
Comment (lencioni) [reply src/DirectionProvider.jsx]: I don't think this actually makes any difference. See here:

image

```jsx
function Bar(props) {
  return <div>{JSON.stringify(props)}</div>;
}

function Foo(props) {
  return <Bar dir={props.dir} lang={props.lang} />;
}

ReactDOM.render(
  <Foo dir="rtl" />,
  document.getElementById('hello-example')
);
```

**Review by lencioni — APPROVED** [review]

File: tests/DirectionProvider_test.jsx [hunk 4]
```diff
@@ -37,6 +37,15 @@ describe('<DirectionProvider>', () => {
     expect(wrapper).to.have.prop('dir', direction);
   });
 
+  it('renders a lang attribute when the lang prop is set', () => {
+    const direction = DIRECTIONS.RTL;
+    const wrapper = shallow(
+      <DirectionProvider direction={direction} lang="ar">{children}</DirectionProvider>,
+    );
+    expect(wrapper).to.have.attribute('lang');
```
Comment (lencioni) [inline tests/DirectionProvider_test.jsx]: Was this line intentional?
```suggestion
```

**Comment by jpnelson** [comment]
> Should there be a propType that restricts it to a valid language code?

It's worth considering, but I think the complete list is pretty substantial (something like 8000+ according to https://r12a.github.io/app-subtags/ when you include subtags) – there's also defined behavior for "invalid" tags in the spec, so I think it's probably safer to just say any string value will be passed on as an attribute.

I think this is also passing CI apart from node v4 now!

File: src/DirectionProvider.jsx (hunk 1 above)
Comment (ljharb) [reply src/DirectionProvider.jsx]: That's because `JSON.stringify` specially removes `undefined` values. The prop is still actually present, so any "exact" propTypes, TS types, or Flow types would error on it.

Comment (jpnelson) [reply src/DirectionProvider.jsx]: Fair enough! I've added that

**Review by ljharb — APPROVED** [review]

**Comment by ljharb** [comment]
Give me a few; i'll get node 4 passing and add that fix to this rebased PR, and then we can land both.

File: src/AutoDirectionProvider.jsx [hunk 5]
```diff
@@ -30,6 +32,7 @@ function AutoDirectionProvider({
     <DirectionProvider
       direction={dir}
       inline={inline}
+      {...(lang && { lang })}
```
Comment (lencioni) [inline src/AutoDirectionProvider.jsx:35]: Why not this?
```suggestion
      lang={lang}
```

File: src/DirectionProvider.jsx [hunk 6]
```diff
@@ -45,10 +46,10 @@ export default class DirectionProvider extends React.Component {
   }
 
   render() {
-    const { children, direction, inline } = this.props;
+    const { children, direction, inline, lang } = this.props;
     const Tag = inline ? 'span' : 'div';
     return (
-      <Tag dir={direction}>
+      <Tag dir={direction} {...(lang && { lang })}>
```
Comment (lencioni) [inline src/DirectionProvider.jsx:52]: Same here?
```suggestion
      <Tag dir={direction} lang={lang}>
```

File: src/AutoDirectionProvider.jsx (hunk 5 above)
Comment (ljharb) [reply src/AutoDirectionProvider.jsx:35]: that would pass `lang={undefined}` when it's not present. i guess that'd be fine tho, unless the `Tag` component used "exact" props and wasn't expecting it. lmk if you want me to hold off merging it pending that change.

Comment (lencioni) [reply src/AutoDirectionProvider.jsx:35]: I don't think it does. See here: https://github.com/airbnb/react-with-direction/pull/29#discussion_r719334819

Comment (lencioni) [reply src/AutoDirectionProvider.jsx:35]: Oh wait, I just saw your response to that. GitHub's formatting of this stuff is hard to follow sometimes.

Comment (lencioni) [reply src/AutoDirectionProvider.jsx:35]: Anyway, it probably doesn't matter much no matter what we do here. But, this would be slightly more efficient for folks using the old JSX transform if there was a single spread without any extra props rather than a mix of props and spreaded props, due to the way the JSX transform works.

Comment (ljharb) [reply src/AutoDirectionProvider.jsx:35]: Since master's tests are broken, maybe let's land this as-is, and then a followup can change the way the lang prop is passed?

(Another alternative is an if/else, where `lang` is explicit in one branch and omitted in the other)

Comment (ljharb) [reply src/AutoDirectionProvider.jsx:35]: I'm going to go ahead and do that; if that's a problem I can put up the followup.
